kernel: add context‑free block validation API (`btck_check_block_context_free`) with POW/Merkle flags

w0xlt opened the pull request:
This PR exposes Bitcoin Core’s context‑free block checks to library users via a new C API entry point, `btck_check_block_context_free`. 

Callers can validate a block’s structure (size/weight, coinbase rules, per‑tx context‑free checks) and optionally re‑run Proof‑of‑Work and Merkle‑root verification without touching chainstate, the block index, or the UTXO set. 

Rationale
Clients embedding the kernel need a pure block sanity check without requiring node state or disk writes (candidate block validation, for example). This API offers that surface in a single call, with optional PoW/Merkle toggles to avoid redundant work when the header has already been validated or when Merkle verification is deferred.

purpleKarrot commented:
Both @w0xlt and @yuvicc, please don't introduce more [boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap)s.

w0xlt committed:
kernel: Expose consensus parameters (`btck_ConsensusParams`)

Library users currently need to maintain a full context object to perform
context-free block validation. Exposing an opaque `btck_ConsensusParams`
struct allows callers to supply only the required consensus parameters,
resulting in a lighter-weight API and a clearer expression of the actual
validation behavior.

Co-authored-by: yuvicc <yuvichh01@gmail.com>

w0xlt committed:
kernel: Expose context-free block validation

This introduces a context-free validation entry point for full blocks in
the kernel C and C++ APIs.

* Add `btck_block_check`, a C function that wraps `CheckBlock` and runs
header and body checks for a `btck_Block` using `btck_ConsensusParams`.
Callers provide a `btck_BlockValidationState` to receive the result
and supply a `btck_BlockCheckFlags` bitmask to control POW and
merkle-root verification.

* Add `btck_BlockCheckFlags` in the C API, plus the corresponding
`BlockCheckFlags` scoped enum in the C++ wrapper, including a
`*_ALL` convenience value.

* Add `Block::Check()` to the C++ wrapper to mirror the new C function
and return a bool while filling a `BlockValidationState`.

* Add a test `(btck_check_block_context_free)` that verifies a known
valid mainnet block passes with `BlockCheckFlags::ALL` and that
truncated block data fails deserialization.

Co-authored-by: yuvicc <yuvichh01@gmail.com>

yuvicc commented on src/kernel/bitcoinkernel_wrapper.h:
```diff
@@ -702,6 +702,39 @@ class BlockHash : public Handle<btck_BlockHash, btck_block_hash_copy, btck_block
+    explicit BlockValidationStateView(const btck_BlockValidationState* ptr) : View{ptr} {}
+};
+
+class BlockValidationState : public UniqueHandle<btck_BlockValidationState, btck_block_validation_state_destroy>, public BlockValidationStateApi<BlockValidationState>
```
And then use `Handle` instead of `UniqueHandle` here:

```suggestion
class BlockValidationState : public Handle<btck_BlockValidationState, btck_block_validation_state_copy, btck_block_validation_state_destroy>, public BlockValidationStateApi<BlockValidationState>
```

yuvicc reviewed (commented):
Approach ACK

Thanks for taking my suggestion.

purpleKarrot commented:
I think the most important question is this: Is there a valid use case for an invalid block?
Could "valid" be an [invariant](https://en.wikipedia.org/wiki/Invariant_(mathematics)#Invariants_in_computer_science) of block, so that a block is guaranteed to be valid by construction?

sedited reviewed (commented):
Concept ACK

w0xlt commented:
Thanks for the ping @yuvicc 
I rebased and addressed all the comments here and in https://github.com/bitcoin/bitcoin/pull/33796 as well.

w0xlt commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1143,6 +1143,33 @@ BITCOINKERNEL_API btck_Block* BITCOINKERNEL_WARN_UNUSED_RESULT btck_block_create
+ *
+ * Runs full context-free block checks (header + body) using the supplied
+ * consensus parameters. The proof-of-work and merkle-root checks can be
+ * toggled via @p flags.
```
Done. Thanks.

w0xlt commented on src/kernel/bitcoinkernel_wrapper.h:
```diff
@@ -702,6 +702,39 @@ class BlockHash : public Handle<btck_BlockHash, btck_block_hash_copy, btck_block
+        return static_cast<const Derived*>(this)->get();
+    }
+    friend Derived;
+public:
```
Done. Thanks.

w0xlt commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1143,6 +1143,33 @@ BITCOINKERNEL_API btck_Block* BITCOINKERNEL_WARN_UNUSED_RESULT btck_block_create
 
+/** Bitflags to control context-free block checks (optional). */
+typedef uint32_t btck_BlockCheckFlags;
+#define btck_BlockCheckFlags_NONE   ((btck_BlockCheckFlags)0)
```
Done. Thanks.

w0xlt commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1077,6 +1077,18 @@ btck_Block* btck_block_copy(const btck_Block* block)
+{
+    auto& state = btck_BlockValidationState::get(validation_state);
+
+    const bool check_pow_again = (flags & btck_BlockCheckFlags_POW) != 0;
```
Done. Thanks.

yuvicc commented on src/kernel/bitcoinkernel.h:
```diff
@@ -853,6 +858,16 @@ BITCOINKERNEL_API btck_ChainParameters* BITCOINKERNEL_WARN_UNUSED_RESULT btck_ch
+ */
+BITCOINKERNEL_API const btck_ConsensusParams* BITCOINKERNEL_WARN_UNUSED_RESULT btck_chain_parameters_get_consensus_params(
+    const btck_ChainParameters* chain_parameters) BITCOINKERNEL_ARG_NONNULL(1);
+
```
doc naming convention: always use struct names instead of natural names.

<details>
  <summary>git diff on d739bb5ae665859cd639a6d0357e8d5dab35a5ed</summary>

```diff
diff --git a/src/kernel/bitcoinkernel.h b/src/kernel/bitcoinkernel.h
index 9bd698fc54..7912e38ef0 100644
--- a/src/kernel/bitcoinkernel.h
+++ b/src/kernel/bitcoinkernel.h
@@ -859,11 +859,11 @@ BITCOINKERNEL_API btck_ChainParameters* BITCOINKERNEL_WARN_UNUSED_RESULT btck_ch
     const btck_ChainParameters* chain_parameters) BITCOINKERNEL_ARG_NONNULL(1);

 /**
- * @brief Get consensus params from chain parameters. The returned consensus params
+ * @brief Get btck_ConsensensusParams from chain parameters. The returned consensus params
  * are not owned and depend on the lifetime of the chain parameters.
  *
  * @param[in] chain_parameters  Non-null.
- * @return                      The consensus parameters.
+ * @return                      The btck_ConsensensusParams.
  */
 BITCOINKERNEL_API const btck_ConsensusParams* BITCOINKERNEL_WARN_UNUSED_RESULT btck_chain_parameters_get_consensus_params(
     const btck_ChainParameters* chain_parameters) BITCOINKERNEL_ARG_NONNULL(1);
```

w0xlt replied:
Done. Thanks.

yuvicc reviewed (commented):
Left some nits.

alexanderwiederin reviewed (commented):
Concept ACK

yuvicc reviewed (commented):
lgtm! ACK f65a692490ab7a0ade8f3623ba1184bea22e187d

Also, wrote an example C code [here](https://github.com/yuvicc/test-bitcoinkernel/blob/master/btck_CheckBlock.c).

alexanderwiederin commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1251,6 +1251,34 @@ BITCOINKERNEL_API btck_Block* BITCOINKERNEL_WARN_UNUSED_RESULT btck_block_create
+/** Bitflags to control context-free block checks (optional). */
+typedef uint32_t btck_BlockCheckFlags;
+#define btck_BlockCheckFlags_BASE   ((btck_BlockCheckFlags)0)
+#define btck_BlockCheckFlags_POW    ((btck_BlockCheckFlags)(1U << 0))  /* run CheckProofOfWork via CheckBlockHeader */
```
I think these comments should be prefixed with `//!<` for doxygen inline comments. What do you think?

alexanderwiederin reviewed (commented):
Thanks for the updates.

w0xlt commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1238,6 +1251,34 @@ BITCOINKERNEL_API btck_Block* BITCOINKERNEL_WARN_UNUSED_RESULT btck_block_create
 
+/** Bitflags to control context-free block checks (optional). */
+typedef uint32_t btck_BlockCheckFlags;
+#define btck_BlockCheckFlags_BASE   ((btck_BlockCheckFlags)0)
```
Done. Thanks.

alexanderwiederin reviewed (commented):
Please add a test for a block that fails the base checks (`btck_BlockCheckFlags_BASE`).

ACK https://github.com/bitcoin/bitcoin/pull/33908/changes/ae742803d5ee0b5e7aec0b142decc9d902a1c849

w0xlt commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1247,6 +1262,38 @@ BITCOINKERNEL_API btck_Block* BITCOINKERNEL_WARN_UNUSED_RESULT btck_block_create
+ *                              optional POW and merkle-root checks. Use
+ *                              btck_BlockCheckFlags_BASE to run only the base
+ *                              checks.
+ * @param[in,out] validation_state Non-null btck_BlockValidationState, will be filled with the result.
```
Done. Thanks.

sedited reviewed (approved):
ACK 9ebe9c7e7b6728b0f57f549d6a6ff7fe71690cee

sedited reviewed (approved):
ACK 0587c56091faf7851b0dd1b6fcbe660ee55f1f10
